tiledb/sm/array_schema/dimension_label_schema.h

```diff
@@ -0,0 +1,174 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * Defines the array schema class
+ */
+
+#ifndef TILEDB_dimension_label_schema_H
+#define TILEDB_dimension_label_schema_H
+
+#include "tiledb/common/common.h"
+#include "tiledb/sm/enums/label_order.h"
+#include "tiledb/sm/enums/layout.h"
+#include "tiledb/sm/misc/constants.h"
+
+using namespace tiledb::common;
+
+namespace tiledb::type {
+class Range;
+}
+
+using namespace tiledb::type;
+
+namespace tiledb::sm {
+
+class ArraySchema;
+class Attribute;
+class Dimension;
+class ByteVecValue;
+class FilterPipeline;
+
+/**
+ * Return a Status_DimensionLabelSchema error class Status with a given message
+ **/
+inline Status Status_DimensionLabelSchemaError(const std::string& msg) {
+  return {"[TileDB::DimensionLabelSchema] Error", msg};
+}
+
+/**
+ * Schema for an dimension label. An dimension label consists of two
+ * one-dimensional arrays used to define a dimension label.
+ */
+class DimensionLabelSchema {
+ public:
+  /**
+   * Size type for the number of labels in an dimension label and for label
+   * indices.
+   *
+   * This must be the same as ArraySchema::attribute_size_type
+   */
+  using attribute_size_type = unsigned int;
+
+  /** Default constructor is not C.41 compliant. */
+  DimensionLabelSchema() = delete;
+
+  /**
+   * Constructor.
+   *
+   * @param label_type The label type.
+   * @param dim The dimension.
+   * @param label_dim_domain The label domain.
+   * @param label_tile_extent The label tile extent.
+   * @param dim_label The dimension label to be created.
+   * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+   */
+  DimensionLabelSchema(
+      LabelOrder label_order,
+      Dimension* dim,
+      Datatype label_type,
+      const void* label_domain,
+      const void* label_tile_extent);
+
+  /**
+   * Constructor.
+   *
+   * @param label_order Order of the labels relative to the index for the
+   * dimension label.
+   * @param indexed_array_schema Array schema for the array with indices defined
+   * on the dimension.
+   * @param labelled_array_schema Array schema for the array with labels defined
+   * on the dimension.
+   * @param label_attr_id The integer ID for the label attribute on the indexed
+   * array.
+   * @param index_attr_id The integer ID for the index attribute on the labelled
+   * array.
+   */
+  DimensionLabelSchema(
+      LabelOrder label_order,
+      shared_ptr<ArraySchema> indexed_array_schema,
+      shared_ptr<ArraySchema> labelled_array_schema,
+      const attribute_size_type label_attr_id = 0,
```

Comment:
Why do we need these?

Reply:
I am assuming we will eventually want the ability to store multiple labels inside a single array. This is to prepare for that future.

Reply:
Yeah, that's a good idea, let's keep them.